### DISCUSSION TITLE
Doubts about doh

### DISCUSSION THREAD

**Discussion by uazo** [bromite/bromite#1904] [General]
I ask you for an opinion on the doh system, my doubt is how much privacy it is.
I don't know about you, but I expected that when a TCP or UDP connection is closed, each new opening towards the same endpoint would be fundamentally uncoupled from the previous one ... well, that's not the case!

it is quite complicated to explain, I hope to be clear, especially in English. I also expect denials (indeed I hope so), because mine is still an opinion.

I started from [Tracking Users across the Web via TLS Session Resumption](https://svs.informatik.uni-hamburg.de/publications/2018/2018-12-06-Sy-ACSAC-Tracking_Users_across_the_Web_via_TLS_Session_Resumption.pdf) because I wanted to understand if bromite is also affected and why tor has decided to disable it by default.

the conclusion is, after weeks to understand how the network service worked (I found it extremely complex) that, thanks to the partition of the `httpcache` made with the `network isolation key`, indeed, although it is possible for a server to trace the navigation at the network level, the its only possibility is to trace only the calls actually made to its endpoint. the tracking is passive and per session (of the browser, i.e. at each restart it starts all over again).
clearly the partitioning must be active: in bromite the [triple keying](https://docs.google.com/document/d/1hm-t6sEn2xR7YN_R1qS1zTWYhyunCK6HoGSrssoQd3A/edit) is active while they are checking the double one after seeing that it affects loading times (and therefore the introduction of the new flag [kForceIsolationInfoFrameOriginToTopLevelFrame](https://bugs.chromium.org/p/chromium/issues/detail?id=1294930))

I explain technically what I understand:

in connections TLS>=1.2 it is possible that the server is enabled to manage the TLS Session Resumption (which remains optional) as a mechanism to allow faster establishment of an encrypted connection by reusing the protocol already established by a previous handshake.
in fact, the resumption is active during the handshake of the connection security protocols (ClientHello/ServerHello message + extensions): tls 1.2 basically uses a random value generated by the client (the session id), while the 1.3 uses a once generated by the server called session ticket. The evolution took place because the sessionid is expected to pass in clear text but also because it forces the server to operate in stateful mode (it must maintain a server-side sessionid database, forcing it to use resources to store them, which is not feasible in servers with few resources or with many accesses) while the 1.3 ticket is composed of server-side encrypted values ​​(and opaque for the client) that allow operation in stateless mode, excellent for servers with few resources or with high loads. furthermore tls 1.3 would allow handshaking with 0-RTT (zero round trip time), using another mechanism (early), which however is disabled by default because it seems to have serious problems of forward secrecy.

in chromium, HTTP connections in ssl are handled with boringssl, which [does not send](https://source.chromium.org/chromium/chromium/src/+/main:third_party/boringssl/src/ssl/handshake_client.cc;l=301;drc=992ef759ae3a9d5604955e41ee7d77b7d005f241;bpv=1;bpt=1) by default sessionid the sessionid on reconnections in TLS 1.2 and verifies that it is the [same](https://source.chromium.org/chromium/chromium/src/+/main:third_party/boringssl/src/ssl/tls13_client.cc;l=111;drc=992ef759ae3a9d5604955e41ee7d77b7d005f241;bpv=1;bpt=1) in the server response. 
boringssl however allows chromium to [manage a cache](https://source.chromium.org/chromium/chromium/src/+/main:net/ssl/ssl_client_session_cache.cc;l=124;drc=992ef759ae3a9d5604955e41ee7d77b7d005f241;bpv=1;bpt=1?q=SSLClientSessionCache::IsExpired&sq=&ss=chromium%2Fchromium%2Fsrc) of a structure called SSL_SESSION which has an incredible timeout of 2h for [TLS 1.2](https://source.chromium.org/chromium/chromium/src/+/main:third_party/boringssl/src/include/openssl/ssl.h;l=1984?q=SSL_DEFAULT_SESSION_TIMEOUT&sq=&ss=chromium%2Fchromium%2Fsrc) and 7 days for [TLS 1.3](https://source.chromium.org/chromium/chromium/src/+/main:third_party/boringssl/src/ssl/ssl_session.cc;l=378;drc=992ef759ae3a9d5604955e41ee7d77b7d005f241;bpv=1;bpt=1)! and those defaults are never changed.

quic is managed more at a higher level by quiche but operates with the cache in the same way both in [tls 1.3 mode](https://source.chromium.org/chromium/chromium/src/+/main:net/third_party/quiche/src/quic/core/tls_client_handshaker.cc;l=123;drc=992ef759ae3a9d5604955e41ee7d77b7d005f241;bpv=1;bpt=1?q=TlsClientHandshaker::CryptoConnect&ss=chromium%2Fchromium%2Fsrc) and in [QUIC_CRYPTO mode](https://source.chromium.org/chromium/chromium/src/+/main:net/third_party/quiche/src/quic/core/crypto/quic_client_session_cache.cc;l=15;bpv=1;bpt=1) which are the [two available handshaking](https://source.chromium.org/chromium/chromium/src/+/main:net/third_party/quiche/src/quic/core/quic_versions.h;l=149;bpv=1;bpt=1?q=PROTOCOL_QUIC_CRYPTO&ss=chromium%2Fchromium%2Fsrc) modes.

strangely, verifying with wireshark I was unable to extract evidence that would allow me to follow this mechanism (perhaps because it is encoded with HKDF but I don't know exactly), but adding LOG() in SSL_session_reused [in the code](https://source.chromium.org/chromium/chromium/src/+/main:net/socket/ssl_client_socket_impl.cc;l=1084;bpv=1;bpt=1?q=DoHandshakeComplete&ss=chromium%2Fchromium%2Fsrc) actually the TLS resumption is active and functioning in chromium and in particular used by quic (THE protocol used by google servers) to resume the (old) connections during the security handshake of the UDP port used. I have also see these on H2 connections and in particular I saw it on DOH calls.

and here it seems to me there is the problem: the dns client used by chromium is in fact not partitioned at the network level, because [it reuses](https://source.chromium.org/chromium/chromium/src/+/main:net/dns/host_resolver_manager.cc;l=667;drc=992ef759ae3a9d5604955e41ee7d77b7d005f241;bpv=1;bpt=1?q=HostResolverManager::RequestImpl&ss=chromium%2Fchromium%2Fsrc) the NetworkIsolationKey of the network service through a [single factory](https://source.chromium.org/chromium/chromium/src/+/main:services/network/network_service.cc;l=335;drc=992ef759ae3a9d5604955e41ee7d77b7d005f241;bpv=1;bpt=1) that produces a single system-wide [ResolveContext](https://source.chromium.org/chromium/chromium/src/+/main:net/dns/resolve_context.cc;l=131;drc=992ef759ae3a9d5604955e41ee7d77b7d005f241;bpv=1;bpt=1)
the partition only covers the responses, ie the dns resolution cache is linked to the httpcache (which is the navigation cache), but not the request, which is always under the same network isolation key at the "system network" level, when instead the user navigation uses different network isolation keys at the level of "profile network" which is partitioned by connection, so it shares the same tcp and udp ports used by the connection protocols (h2 or h3).
the reply of the dns, on the other hand, is however inserted in the correct httpcache through the isolation info, which operate at the frame level but can use different network isolation (also shared, such as those of the system network).

therefore, the tls resumption allows (perhaps) the doh server to trace non-consecutive requests, associating them to the same device, with persistent values between different connections, so even if the connection is closed and reopened later, until the application is restarted. in fact, the cache is only in memory, although the code [allow saving to disk](https://source.chromium.org/chromium/chromium/src/+/main:net/quic/quic_stream_factory.cc;l=1816;drc=992ef759ae3a9d5604955e41ee7d77b7d005f241;bpv=1;bpt=1?q=QuicStreamFactory::CreateSession&ss=chromium%2Fchromium%2Fsrc) but it is disabled by default (`max_server_configs_stored_in_properties` == 0).

among other things, and I have tried, the deactivation of the tls resumption is not the only problem: you can also check it using [PCAPdroid](https://github.com/emanuele-f/PCAPdroid).
the connection to the doh server remains open until the timeout, and is more clearly visible if you use the cloudflare services. moreover it is also active in incognito mode and despite not having active cookies and javascript, being at the network level and not at the profile.
in fact the other issue is that H2 connections (in TLS) have two client-side timeout stages, both set by default at [5 minutes](https://source.chromium.org/chromium/chromium/src/+/main:net/socket/client_socket_pool.cc;l=36;drc=992ef759ae3a9d5604955e41ee7d77b7d005f241;bpv=1;bpt=1?q=g_used_idle_socket_timeout_s&sq=&ss=chromium%2Fchromium%2Fsrc): the first timeout brings the connection to idle, but it is not closed permanently until the [second timeout](https://source.chromium.org/chromium/chromium/src/+/main:net/socket/transport_client_socket_pool.cc;l=898;drc=992ef759ae3a9d5604955e41ee7d77b7d005f241;bpv=1;bpt=1?q=CleanupIdleSocketsInGroup&ss=chromium%2Fchromium%2Fsrc), allowing it to be reused. QUIC connections (udp), on the other hand, are not reusable, they are closed at the first timeout, but the TLS session resumption has the default timeout set at days.
Among the DOH services present in chromium, the latter I have seen active only with dns.google, while the H2 is active in the others but strangely I have noticed different behaviors perhaps because they have different timeouts on the server side.

now, my doubt is, is it a privacy issue? if I read the privacy policies of the doh services selectable in chromium they say that they do not keep the ip, but only some logs at the network level, without specifying which data. it is also true that, even without the tls resumption, using their services still means sending them the navigation, and this is clearly evident and specified in the policies and despite being aware of it, I honestly did not think it was so easy to detect the entire history of my device, at least until the application is restarted. and there is no cancellation that holds because the system network isolation key is not reset except at startup (in fact, the cancellation of the navigation data should always correspond to a restart, but it is not) and the SSL_SESSION cache is only deleted if the passwords are also deleted ...

**Comment by csagan5** [comment]
> I expected that when a TCP or UDP connection is closed, each new opening towards the same endpoint would be fundamentally uncoupled from the previous one

Why? Have you read it somewhere in a specification for DNS-over-HTTPS?

> now, my doubt is, is it a privacy issue?

The only declared purpose of DoH (similarly to DNS-over-TLS and DNS-over-QUIC, see https://dnsprivacy.org/the_solutions/) is to move most of the users' DNS traffic to an encrypted form instead of cleartext over UDP (regular DNS over port 53), and possibly obfuscate it as HTTPS traffic. You still trust your DNS server with your IP and queries, creating new connections every time would not change anything in terms of privacy. In that dnsprivacy.org page there are links about the specs and what they specify (or not).

**Reply by uazo** [reply]
> Why? Have you read it somewhere in a specification for DNS-over-HTTPS?

No, in fact no

> You still trust your DNS server with your IP and queries, 

perfect, my doubt was indeed, you have to use your own server, basically.

> creating new connections every time would not change anything in terms of privacy

do you also say that it is not worth mitigating the problem?

**Reply by csagan5** [reply]
> QUIC connections (udp), on the other hand, are not reusable, they are closed at the first timeout

There's nothing to close with UDP?

> do you also say that it is not worth mitigating the problem?

What problem? You mentioned 2-3 but I fail to see how any of them applies to the interaction with a DoH server e.g. what extra information is the server gaining?

Perhaps these could be problem if you connect to VPN and thus server-side can track you even when changing IP, but that has nothing to do with DoH and everything to do with the networking layer.

**Reply by uazo** [reply]
> There's nothing to close with UDP?

I do not understand

> You mentioned 2-3 but I fail to see how any of them applies to the interaction with a DoH server e.g. what extra information is the server gaining?

I think the main problem is the ease of connecting successive requests per device

> Perhaps these could be problem if you connect to VPN and thus server-side can track you even when changing IP

I can check it out

> but that has nothing to do with DoH and everything to do with the networking layer.

ok, but in bromite the system network layer is used only for doh, checking updates and download ads filters. if it is used for anything else it is probably a bug that should be blocked.

**Reply by csagan5** [reply]
> > There's nothing to close with UDP?
> 
> I do not understand

With TCP you have connection state, with UDP no. There are just packets back and forth, no "open", "half-open", "closed" states.

> > You mentioned 2-3 but I fail to see how any of them applies to the interaction with a DoH server e.g. what extra information is the server gaining?
> 
> I think the main problem is the ease of connecting successive requests per device

Why is that a problem? Requests are coming from the same IP address, you can group them already by that.

> > but that has nothing to do with DoH and everything to do with the networking layer.
> 
> ok, but in bromite the system network layer is used only for doh, checking updates and download ads filters. if it is used for anything else it is probably a bug that should be blocked.

Are you referring to session state shared due to TLS state shared? If there is a bug and you can make a testcase you should submit it as a security issue upstream.

**Reply by uazo** [reply]
> With TCP you have connection state, with UDP no

are you referring to quic? quic is a protocol developed on top of udp, which handles connections and flow control like an tcp connection 

this is an excerpt from the protocol description, but the complete protocol is much more complex:

> QUIC connections are identified by a 64 bit connection ID, randomly generated by the client. In contrast, TCP connections are identified by a 4-tuple of source address, source port, destination address and destination port. This means that if a client changes IP addresses (for example, by moving out of Wi-Fi range and switching over to cellular) or ports (if a NAT box loses and rebinds the port association), any active TCP connections are no longer valid. When a QUIC client changes IP addresses, it can continue to use the old connection ID from the new IP address without interrupting any in-flight requests.

> Requests are coming from the same IP address, you can group them already by that.

sure, but there is a difference between aggregating by ip and by device

> If there is a bug and you can make a testcase you should submit it as a security issue upstream.

ok, I understand

**Reply by uazo** [reply]
> If there is a bug and you can make a testcase you should submit it as a security issue upstream.

regarding the dns, it is already present:
https://bugs.chromium.org/p/chromium/issues/detail?id=1082311
https://bugs.chromium.org/p/chromium/issues/detail?id=997049 blocking https://bugs.chromium.org/p/chromium/issues/detail?id=993801

I can think that the collateral bug of SSL Session resumption allowed in DOH requests in incognito mode is related to the fact that the points above are not yet closed.
regarding the expiration of the ssl sessions I have not found anything (but I do not think they would treat it).

**Reply by uazo** [reply]
in reference to https://bugs.chromium.org/p/chromium/issues/detail?id=1082311

> 3)  The same as 2), but figure out some way to vend new NIKs based on the original NIK, so that normal HTTP requests and DoH requests used different sockets.  This adds some complexity to NIKs, but solves at least one issue with 2).

i am developing this version

**Reply by csagan5** [reply]
> > With TCP you have connection state, with UDP no
> 
> are you referring to quic? quic is a protocol developed on top of udp, which handles connections and flow control like an tcp connection
> 
> this is an excerpt from the protocol description, but the complete protocol is much more complex:

No, I was referring to TCP.

> > Requests are coming from the same IP address, you can group them already by that.
> 
> sure, but there is a difference between aggregating by ip and by device

When talking about security you need to first qualify what is the attack scenario, otherwise no progress can be made.

> > If there is a bug and you can make a testcase you should submit it as a security issue upstream.
> 
> regarding the dns, it is already present: https://bugs.chromium.org/p/chromium/issues/detail?id=1082311 https://bugs.chromium.org/p/chromium/issues/detail?id=997049 blocking https://bugs.chromium.org/p/chromium/issues/detail?id=993801
> 
> I can think that the collateral bug of SSL Session resumption allowed in DOH requests in incognito mode is related to the fact that the points above are not yet closed. regarding the expiration of the ssl sessions I have not found anything (but I do not think they would treat it).

Well, it's swiss cheese then :) if you can describe the attack scenario in more detail we can document it as a known issue and it will be of great help for both users and developers that might want to contribute patches.

> in reference to https://bugs.chromium.org/p/chromium/issues/detail?id=1082311
> 
> > 3. The same as 2), but figure out some way to vend new NIKs based on the original NIK, so that normal HTTP requests and DoH requests used different sockets.  This adds some complexity to NIKs, but solves at least one issue with 2).
> 
> i am developing this version

Sounds good.

**Reply by uazo** [reply]
>> i am developing this version
>
> Sounds good.

here you find the two patches:

https://github.com/uazo/bromite/blob/master-v100/build/patches/contrib/Partition-dns-requests-by-topframe-nik.patch
https://github.com/uazo/bromite/blob/master-v100/build/patches/contrib/Disable-tls-resumption.patch

I have been trying them for a few weeks and have not found any problems. tell me if you want me to do a pull to analyze them in detail

**Reply by csagan5** [reply]
Thanks; please submit 2 PRs so that we can discuss and if they get merged you get attribution on GitHub too.

**Reply by uazo** [reply]
> so that we can discuss

ok, i'll do

> you get attribution on GitHub too

perfect, but you know that I personally don't care...

**Comment by ChandraNakka** [comment]


**Reply by uazo** [reply]
> How can I contact you?

I am here, no private discussion of bromite

**Reply by uazo** [reply]
> I want to contact you regarding my other browser NEON.

thanks but I'm not interested in participating
